pkg/registry/core/service/ipallocator/allocator.go

```diff
@@ -86,12 +86,12 @@ func NewAllocatorCIDRRange(cidr *net.IPNet, allocatorFactory allocator.Allocator
 	base := utilnet.BigForIP(cidr.IP)
 	rangeSpec := cidr.String()
 
-	if utilnet.IsIPv6CIDR(cidr) {
-		// Limit the max size, since the allocator keeps a bitmap of that size.
-		if max > 65536 {
-			max = 65536
-		}
-	} else {
+	// Limit the max size, since the allocator keeps a bitmap of that size in etcd.
+	// Kubernetes Scalability only tests O(10K) services
+	// https://github.com/kubernetes/kubernetes/issues/94029
+	if max > 65536 {
```

Comment:
I'm wondering if there isn't a risk to break some existing users with it?
@thockin @bowei - WDYT?

[In general I'm supportive to make that explicit, but I'm not fully sure how safe it is]

Reply:
We can break users that are already using IPs out of the lower 16 bits, however, those users are the ones that will have etcd database sizes with ~2GB on bitmaps just because of the revisions stored in etcd... 

I don have clear how the upgrade will work here though ... the allocators are created by the apiserver and the repair loop keeps them up to date :thinking:

Reply:
> however, those users are the ones that will have etcd database sizes with ~2GB on bitmaps just because of the revisions stored in etcd...

This isn't true. Compaction is happening every 5m by default, so if you are not creating hunderds/thousands of services in short periods of time, your fine. E.g. if you create 1 service per-hour, you won't face the problem you faced at all. And I wouldn't exclude the sitatuion that it may work with more than 65k services.

Reply:
ic, I didn't count on that, then it is not as problematic as I thought